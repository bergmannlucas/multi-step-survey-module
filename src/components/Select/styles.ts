import styled, { css } from 'styled-components'

export const Wrapper = styled.div``

export const Select = styled.select`
  ${({ theme }) => css`
    position: relative;

    width: 100%;
    max-width: 15rem;

    cursor: pointer;
    color: ${theme.colors.white};
    background: ${theme.colors.purple};
    font-size: ${theme.font.sizes.xlarge};
    font-family: ${theme.font.family};

    text-align: center;
    border-radius: ${theme.border.radius.small};
  `}
`

export const Label = styled.label`
  ${({ theme }) => css`
    margin: 0 1rem 0 0;
    font-size: ${theme.font.sizes.xlarge};
    font-weight: ${theme.font.bold};
    color: ${theme.colors.purple};
  `}
`
