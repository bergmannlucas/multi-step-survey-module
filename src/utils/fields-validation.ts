import Joi from 'joi'

export const schema = Joi.array().items(
  Joi.object({
    name: Joi.string().empty(''),
    email: Joi.string()
      .email({ tlds: { allow: false } })
      .empty('')
  }),
  Joi.object({
    gender: Joi.string().required(),
    age: Joi.number().required()
  }),
  Joi.object({
    book: Joi.string().required(),
    colors: Joi.object({
      black: Joi.boolean(),
      white: Joi.boolean(),
      red: Joi.boolean(),
      green: Joi.boolean(),
      blue: Joi.boolean(),
      yellow: Joi.boolean(),
      purple: Joi.boolean(),
      orange: Joi.boolean(),
      pink: Joi.boolean()
    })
  })
)
