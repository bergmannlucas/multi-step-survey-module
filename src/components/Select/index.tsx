import { SelectHTMLAttributes } from 'react'

import * as S from './styles'

type SelectProps = {
  options?: number[] | string[]
  label?: string
} & SelectHTMLAttributes<HTMLSelectElement>

function Select({ options = [], label, ...props }: SelectProps) {
  return (
    <S.Wrapper>
      {!!label && <S.Label>{label}</S.Label>}
      <S.Select {...props}>
        {options.map((option) => {
          return <option key={option}>{option}</option>
        })}
      </S.Select>
    </S.Wrapper>
  )
}

export default Select
