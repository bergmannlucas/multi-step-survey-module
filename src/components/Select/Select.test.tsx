import userEvent from '@testing-library/user-event'
import { render, screen, waitFor } from 'utils/test-utils'
import Select from '.'

describe('<Select />', () => {
  it('should render with label', () => {
    const { container } = render(
      <Select options={[0, 1, 2]} label="age" onChange={jest.fn()} />
    )

    expect(screen.getByRole('combobox')).toBeInTheDocument()
    expect(screen.getByText(/age/i)).toBeInTheDocument()

    expect(container.firstChild).toMatchSnapshot()
  })

  it('should render without label', () => {
    render(<Select options={[0, 1, 2]} onChange={jest.fn()} />)

    expect(screen.getByRole('combobox')).toBeInTheDocument()
    expect(screen.queryByText(/age/i)).not.toBeInTheDocument()
  })

  it('should dispatch onChange when user selects an option', async () => {
    const onChange = jest.fn()
    render(<Select options={['0', '1', '2']} onChange={onChange} />)

    expect(onChange).not.toHaveBeenCalled()

    userEvent.selectOptions(screen.getByRole('combobox'), '1')
    await waitFor(() => {
      expect(onChange).toHaveBeenCalledTimes(1)
    })
  })
})
