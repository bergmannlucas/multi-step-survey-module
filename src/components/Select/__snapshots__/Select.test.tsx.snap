// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Select /> should render with label 1`] = `
.c1 {
  position: relative;
  width: 100%;
  max-width: 15rem;
  cursor: pointer;
  color: #FAFAFA;
  background: #9249FF;
  font-size: 2.0rem;
  font-family: Montserrat,-apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,Oxygen,Ubuntu,Cantarell,'Open Sans','Helvetica Neue',sans-serif;
  text-align: center;
  border-radius: 0.5rem;
}

.c0 {
  margin: 0 1rem 0 0;
  font-size: 2.0rem;
  font-weight: 700;
  color: #9249FF;
}

<div
  class=""
>
  <label
    class="c0"
  >
    age
  </label>
  <select
    class="c1"
  >
    <option>
      0
    </option>
    <option>
      1
    </option>
    <option>
      2
    </option>
  </select>
</div>
`;
